app/models.py

```diff
@@ -2551,3 +2551,90 @@ def serialize(self):
             "page_count": self.page_count,
         }
+
+
+class UnsubscribeRequestReport(db.Model):
+    __tablename__ = "unsubscribe_request_report"
+    """
+    CREATE VIEW unsubscribe_request_report AS
+    SELECT
+        service_id,
+        unsubscribe_request_report_id as id,
+        min(created_at) as earliest_timestamp,
+        max(created_at) as latest_timestamp,
+        (case when downloaded_at is null then false else true end) as has_been_downloaded
+        count(*)
+    FROM unsubscribe_requests
+    group by service_id, unsubscribe_request_report_id
+    """
+
+    # Tell alembic not to create this as a table. We have a migration where we manually set this up as a view.
+    # This is custom logic we apply - not built-in logic. See `migrations/env.py`
+    __table_args__ = {"info": {"managed_by_alembic": False}}
+
+    id = db.Column(UUID(as_uuid=True), primary_key=True)
+    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey("services.id"), nullable=False)
+    service = db.relationship(Service, backref=db.backref("unsubscribe_requests"))
+
+    earliest_timestamp = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)
+    latest_timestamp = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)
+    has_been_downloaded = db.Column(db.Boolean, nullable=False)
+    count = db.Column(db.Integer, nullable=False)
+
+
+class UnsubscribeRequest(db.Model):
+    __tablename__ = "unsubscribe_request"
+
+    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
+
+    # WIP:
+    # Ignoring a strict foreign key relationship here for now. Notifications are archived to the NotificationHistory
+    # table by a nightly job and I haven't investigated whether that might break a strict FK yet or if it would
+    # work smoothly. We can still have a relationship using an explicit join condition.
+    notification_id = db.Column(UUID(as_uuid=True), index=True, nullable=False)
+    notification = db.relationship(
+        "NotificationAllTimeView",
+        primaryjoin="NotificationLetterDespatch.notification_id == foreign(NotificationAllTimeView.id)",
+        uselist=False,
+    )
+
+    # this is denormalised but might still be useful to have as a separate column?
+    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey("services.id"), nullable=False)
+    service = db.relationship(Service, backref=db.backref("unsubscribe_requests"))
+
+    # is it useful to
+    template_id = db.Column(UUID(as_uuid=True), db.ForeignKey("templates.id"), nullable=False)
+    template_version = db.Column(db.Integer)
+    template = db.relationship(Template, backref=db.backref("unsubscribe_requests"))
+    template_history = db.relationship(TemplateHistory, backref=db.backref("unsubscribe_requests"))
+
+    # nullable as at some point we may want to delete these for data retention purposes
+    email_address = db.Column(db.Text, nullable=True)
+
```

Comment:
I assume that will be dictated by the service's data retention policy?
Since the best practice for one click unsubscribe uri is to have an opaque identifier that doesn't expose PII, could we have for instance an `opaque_identifier` which would be a hash of the email address of some sort?

Reply:
need to check with rest of team this afternoon to answer this.

unsure about data retention. in chris's slides it refers to deleting processed reports after 7 days. we need to decide what that means (do we keep the rows in the db? do we delete the email address PII from the database? do we just hide from the UI? etc)

Reply:
we couldn't have a hash though, because when the user downloads the report they need the original user email address - so we need it to be either be plain text, or two-way encrypted rather than hashed